backend/src/main/java/ai/verta/modeldb/versioning/RepositoryDAORdbImpl.java

```diff
@@ -1312,88 +1310,69 @@ public DatasetPaginationDTO findDatasets(
-        }
+            accessibleAllWorkspaceItems.stream()
+                .map(GetResourcesResponseItem::getResourceId)
+                .collect(Collectors.toSet());
 
-        List<String> orgIds =
+        List<String> orgWorkspaceIds =
             roleService.listMyOrganizations().stream()
-                .map(Organization::getId)
+                .map(Organization::getWorkspaceId)
                 .collect(Collectors.toList());
-        if (!orgIds.isEmpty()) {
-          finalPredicatesList.add(
-              builder.not(
-                  builder.and(
-                      repositoryRoot.get(ModelDBConstants.WORKSPACE_ID).in(orgIds),
-                      builder.equal(
-                          repositoryRoot.get(ModelDBConstants.WORKSPACE_TYPE),
-                          WorkspaceType.ORGANIZATION_VALUE))));
-        }
-      } else {
-        if (resourceVisibility.equals(ResourceVisibility.PRIVATE)) {
-          List<KeyValueQuery> workspacePredicates =
-              ModelDBUtils.getKeyValueQueriesByWorkspace(
-                  roleService, currentLoginUserInfo, workspaceName);
-          if (workspacePredicates.size() > 0) {
-            Predicate privateWorkspacePredicate =
-                builder.equal(
-                    repositoryRoot.get(ModelDBConstants.WORKSPACE_ID),
-                    workspacePredicates.get(0).getValue().getStringValue());
-            Predicate privateWorkspaceTypePredicate =
-                builder.equal(
-                    repositoryRoot.get(ModelDBConstants.WORKSPACE_TYPE),
-                    workspacePredicates.get(1).getValue().getNumberValue());
-            Predicate privatePredicate =
-                builder.and(privateWorkspacePredicate, privateWorkspaceTypePredicate);
-
-            finalPredicatesList.add(privatePredicate);
+        for (GetResourcesResponseItem item : accessibleAllWorkspaceItems) {
+          if (orgWorkspaceIds.contains(String.valueOf(item.getWorkspaceId()))) {
+            accessibleDatasetIds.remove(item.getResourceId());
           }
         }
+      } else {
+        accessibleDatasetIds =
+            ModelDBUtils.filterWorkspaceOnlyAccessibleIds(
+                roleService,
+                !queryParameters.getDatasetIdsList().isEmpty()
+                    ? new HashSet<>(queryParameters.getDatasetIdsList())
+                    : Collections.emptySet(),
+                workspaceName,
+                currentLoginUserInfo,
+                ModelDBServiceResourceTypes.REPOSITORY);
+      }
+
+      if (accessibleDatasetIds.isEmpty() && roleService.IsImplemented()) {
+        LOGGER.debug("Accessible Dataset Ids not found, size 0");
+        DatasetPaginationDTO emptyPaginationDTO = new DatasetPaginationDTO();
+        emptyPaginationDTO.setDatasets(Collections.emptyList());
+        emptyPaginationDTO.setRepositories(Collections.emptyList());
+        emptyPaginationDTO.setTotalRecords(0L);
+        return emptyPaginationDTO;
```

Comment:
This code is duplicated a few times, please extract to a private utility method.

Reply:
Extracted